controllers/nodehealthcheck_controller.go

```diff
@@ -318,47 +317,88 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	// This happens e.g. on cloud providers with Machine Deletion remediation: the broken node will be deleted and
 	// a new node created, with a new name, and no relationship to the old node
-	if err = r.deleteOrphanedRemediationCRs(nhc, append(healthyNodes, unhealthyNodes...), resourceManager, log); err != nil {
+	if err = r.deleteOrphanedRemediationCRs(nhc, append(notMatchingNodes, matchingNodes...), resourceManager, log); err != nil {
 		return result, err
 	}
 
 	// delete remediation CRs for healthy nodes
-	for _, node := range healthyNodes {
+	healthyCount := 0
+	for _, node := range notMatchingNodes {
 		log.Info("handling healthy node", "node", node.GetName())
 		remediationCRs, err := resourceManager.ListRemediationCRs(nhc, func(cr unstructured.Unstructured) bool {
-			return cr.GetName() == node.GetName()
+			return cr.GetName() == node.GetName() && resources.IsOwner(&cr, nhc)
 		})
 		if err != nil {
 			log.Error(err, "failed to get remediation CRs for healthy node", "node", node.Name)
 			return result, err
 		}
-		for _, remediationCR := range remediationCRs {
-			if err := r.deleteRemediationCR(&remediationCR, nhc, resourceManager, log); err != nil {
-				log.Error(err, "failed to delete remediation CR", "name", remediationCR.GetName())
-				return result, err
+
+		if len(remediationCRs) == 0 {
+			// when all CRs are gone, the node is considered healthy
+			resources.UpdateStatusNodeHealthy(node.GetName(), nhc, r.Recorder)
+			healthyCount++
+			continue
+		}
+
+		// set conditions healthy timestamp
+		conditionsHealthyTimestamp := resources.UpdateStatusNodeConditionsHealthy(&node, nhc, currentTime())
+		if conditionsHealthyTimestamp != nil {
+			// warn about pending CRs when all CRs have been deleted for some time already but still exist
+			doLog := true
+			logThreshold := conditionsHealthyTimestamp.Add(-logWhenCRPendingDeletionDuration)
+			for _, cr := range remediationCRs {
+				if cr.GetDeletionTimestamp() == nil || cr.GetDeletionTimestamp().After(logThreshold) {
+					doLog = false
+					// requeue when we need to log
+					var logIn time.Duration
+					if cr.GetDeletionTimestamp() != nil {
+						logIn = cr.GetDeletionTimestamp().Sub(logThreshold) + time.Second
+					} else {
+						logIn = logWhenCRPendingDeletionDuration + time.Second
+					}
+					finalRequeueAfter = utils.MinRequeueDuration(finalRequeueAfter, &logIn)
+				}
+			}
+			if doLog {
+				log.Info("Node conditions don't match unhealthy condition anymore, but node has remediation CR(s) with pending deletion, considering node as unhealthy")
 			}
 		}
+
+		if err = r.deleteRemediationCRs(remediationCRs, nhc, resourceManager, false, log); err != nil {
+			return result, err
+		}
 	}
 
+	nhc.Status.ObservedNodes = pointer.Int(len(selectedNodes))
+	nhc.Status.HealthyNodes = &healthyCount
+
 	// we are done in case we don't have unhealthy nodes
-	if len(unhealthyNodes) == 0 {
+	if len(matchingNodes) == 0 {
```

Comment:
Does it make sense to also check here whether the value of `healthyCount` is greater than zero?

Reply:
I don't think so, what would be the value? 🤔

Reply:
IIUC the amount of healthy nodes is greater than zero, since the total number of selected nodes is their sum (len(matchingNodes) + healthyCount= NUMBER_OF_SELCTED_NODES). Therefore, when `len(matchingNodes) == 0 ` then `healthyCount` should be greater than zero (and equal to the number of selected cluster nodes based on NHC selector).

 But I might be wrong here when the amount of selected nodes is zero (if it is possible)

Reply:
Yes, there should be healthy nodes, but it just doesn't matter at this place. The following code is going to handle unhealthy nodes. If there aren't any unhealthy nodes, we can skip all of that. Healthy nodes are uninteresting here.